src/Compilers/CSharp/Portable/Binder/ExpressionVariableFinder.cs

```diff
@@ -274,15 +246,97 @@ public override void VisitDeclarationExpression(DeclarationExpressionSyntax node
+            {
+                _localsBuilder.Add(variable);
+            }
+        }
+
+        protected abstract TFieldOrLocalSymbol MakeOutVariable(DeclarationExpressionSyntax node, BaseArgumentListSyntax argumentListSyntax, SyntaxNode nodeToBind);
+    }
+
+    internal class ExpressionVariableFinder : ExpressionVariableFinder<LocalSymbol>
+    {
+        private Binder _scopeBinder;
+        private Binder _enclosingBinder;
+
+
+        internal static void FindExpressionVariables(
+            Binder scopeBinder,
+            ArrayBuilder<LocalSymbol> builder,
+            CSharpSyntaxNode node,
+            Binder enclosingBinderOpt = null)
+        {
+            if (node == null)
+            {
+                return;
+            }
+
+            var finder = s_poolInstance.Allocate();
+            finder._scopeBinder = scopeBinder;
+            finder._enclosingBinder = enclosingBinderOpt ?? scopeBinder;
+
+            finder.FindExpressionVariables(builder, node);
+
+            finder._scopeBinder = null;
+            finder._enclosingBinder = null;
+            s_poolInstance.Free(finder);
+        }
+
+        internal static void FindExpressionVariables(
+            Binder binder,
+            ArrayBuilder<LocalSymbol> builder,
+            SeparatedSyntaxList<ExpressionSyntax> nodes)
+        {
+            if (nodes.Count == 0)
+            {
+                return;
+            }
+
+            var finder = s_poolInstance.Allocate();
+            finder._scopeBinder = binder;
+            finder._enclosingBinder = binder;
+
+            finder.FindExpressionVariables(builder, nodes);
+
+            finder._scopeBinder = null;
+            finder._enclosingBinder = null;
+            s_poolInstance.Free(finder);
+        }
+
+        protected override LocalSymbol MakePatternVariable(DeclarationPatternSyntax node, SyntaxNode nodeToBind)
+        {
+            return SourceLocalSymbol.MakeLocalSymbolWithEnclosingContext(
+                            _scopeBinder.ContainingMemberOrLambda,
+                            scopeBinder: _scopeBinder,
+                            nodeBinder: _enclosingBinder,
+                            typeSyntax: node.Type,
+                            identifierToken: node.Identifier,
+                            kind: LocalDeclarationKind.PatternVariable,
+                            nodeToBind: nodeToBind,
+                            forbiddenZone: null);
+        }
+        protected override LocalSymbol MakeOutVariable(DeclarationExpressionSyntax node, BaseArgumentListSyntax argumentListSyntax, SyntaxNode nodeToBind)
+        {
+            NamedTypeSymbol container = _scopeBinder.ContainingType;
+
+            if ((object)container != null && container.IsScriptClass && 
+                (object)_scopeBinder.LookupDeclaredField(node.VariableDesignation()) != null)
+            {
+                // This is a field declaration
+                return null;
```

Comment:
>null [](start = 23, length = 4)

It seems that this condition could be incorrectly satisfied inside a top-level method, or in a lambda.